### PR TITLE
workload/schemachange: update and disable `ALTER TYPE DROP VALUE`

### PR DESCRIPTION
**Pull request by chrisseto**
#### 31a2ac9d7d2576d9d0761b0588118da93ae49564 workload/schemachange: update and disable `ALTER TYPE DROP VALUE`

Previously, `ALTER TYPE DROP VALUE` was using the procedural generation
that is most common in this workload.

The desire to add support UDFs required revisiting the implement as
referenced enum values cannot be dropped.

Unfortunately, there is not a straightforward way to determine if a
specific enum is being referenced by other descriptors (namely default
expressions for columns and UDFs).

After the initial update and a few testing rounds with UDF support, a
bug between enums and UDFs was discovered. (See #114844 for details).

This commit upgrades the implementation of `ALTER TYPE DROP VALUE` to
utilize the `Generate` helper and dodge cases of referenced enum values.

Due to the aforementioned bug and another outstanding flake, this commit
also disabled `ALTER TYPE DROP VALUE` to prevent further flakes an
noise.

Epic: CRDB-19168
Informs: #113859, #114844
Release note: None

### CODE REVIEW

File: pkg/workload/schemachange/optype.go [hunk 1]
```diff
@@ -275,7 +275,7 @@ var opWeights = []int{
 	dropSequence:                      1,
 	dropTable:                         1,
 	dropView:                          1,
-	alterTypeDropValue:                1,
+	alterTypeDropValue:                0, // Disabled and tracked with #114844 and #113859
```
Comment (Xiang-Gu) [inline pkg/workload/schemachange/optype.go:278]: Let's also add #115612 as one of the tracking issue in the comment as well

File: pkg/workload/schemachange/query_util.go [hunk 2]
```diff
@@ -18,9 +18,38 @@ import (
...
+	//
+	// id::int | schema_id::int | name::text | descriptor::json
+	descJSONQuery = `SELECT
+		descriptor.id,
+		"parentSchemaID" AS schema_id,
+		namespace.name AS name,
+		crdb_internal.pb_to_json('desc', descriptor) AS descriptor
+	FROM system.descriptor
+	JOIN system.namespace ON namespace.id = descriptor.id
+	WHERE "parentID" = (SELECT id FROM system.namespace WHERE name = current_database() AND "parentID" = 0)
```
Comment (Xiang-Gu) [inline pkg/workload/schemachange/query_util.go:33]: why do we have the `AND "parentID" = 0` here?

In other words, why isn't 
```
SELECT id FROM system.namespace WHERE name = current_database();
```
sufficient to find the ID of the current database?

File: pkg/workload/schemachange/query_util.go [hunk 3]
```diff
@@ -18,9 +18,38 @@ import (
...
+	`
+
+	// enumDescsQuery returns the JSONified version of all enum descriptors in
+	// the current database.
+	//
+	// [descJSONQuery] must be bound to the name "descriptors".
+	//
+	// id::int | schema_id::int | name::text | descriptor::json
+	enumDescsQuery = `SELECT id, schema_id, name, descriptor->'type' AS descriptor FROM descriptors WHERE descriptor ? 'type'`
```
Comment (Xiang-Gu) [inline pkg/workload/schemachange/query_util.go:42]: what does "WHERE descriptor ? 'type'" mean here? Tried some online searching but didn't find a definite answer (is it a placeholder? If so, what value is substituted later for this placeholder?)

File: pkg/workload/schemachange/operation_generator.go [hunk 4]
```diff
@@ -2021,37 +2021,53 @@ func (og *operationGenerator) dropView(ctx context.Context, tx pgx.Tx) (*opStmt,
 	return stmt, nil
 }
 
-func (og *operationGenerator) dropTypeValue(ctx context.Context, tx pgx.Tx) (*opStmt, error) {
-	enum, exists, err := og.randEnum(ctx, tx, og.pctExisting(true))
+func (og *operationGenerator) alterTypeDropValue(ctx context.Context, tx pgx.Tx) (*opStmt, error) {
+	query := With([]CTE{
```
Comment (Xiang-Gu) [inline pkg/workload/schemachange/operation_generator.go:2025]: can we add some comment here about what this query tries to do?

File: pkg/workload/schemachange/operation_generator.go [hunk 5]
```diff
@@ -2021,37 +2021,53 @@ func (og *operationGenerator) dropView(ctx context.Context, tx pgx.Tx) (*opStmt,
...
+	// referenced. Unreferenced members can be dropped but referenced members may
+	// not. For now, we skip over all enums where the type itself is being
+	// referenced.
+
+	stmt, code, err := Generate[*tree.AlterType](og.params.rng, og.produceError(), []GenerationCase{
+		// Fail to drop values from a type that doesn't exist.
+		{pgcode.UndefinedObject, `ALTER TYPE "EnumThatDoesntExist" DROP VALUE 'IrrelevantValue'`},
+		// Fail to drop a value that is in the process of being dropped.
+		{pgcode.ObjectNotInPrerequisiteState, `{ with (EnumValue true false) } ALTER TYPE { .name } DROP VALUE { .value } { end }`},
+		// Fail to drop types that don't exist.
```
Comment (Xiang-Gu) [inline pkg/workload/schemachange/operation_generator.go:2053]: nit: should it be "fail to drop a value that does not exit?"

**Review by Xiang-Gu — APPROVED** [review]
Thanks for doing this. I left a few comments. I'm okay disabling this for now (but I'm also curious to see if https://github.com/cockroachdb/cockroach/issues/115612 would fix anything, but it should not block us from merging this first). We can always enable it again later and see once that issue is fixed.

File: pkg/workload/schemachange/query_util.go (hunk 2 above)
Comment (chrisseto) [reply pkg/workload/schemachange/query_util.go:33]: `"parentID" = 0` ensures that we're selecting a database form the namespace table. If we omitted that and someone created, say, a table `movr` in `cockroach demo`, the query you've posted will return multiple rows.

File: pkg/workload/schemachange/operation_generator.go (hunk 4 above)
Comment (chrisseto) [reply pkg/workload/schemachange/operation_generator.go:2031]: done!

**Comment by chrisseto** [comment]
bors r=Xiang-Gu

**Comment by rafiss** [comment]
blathers backport 23.2.0-rc 23.2

**Comment by chrisseto** [comment]
Ah, I'm guessing the conflicts are because I alphabetized some of the operations. I'll take a stab at manually backporting!

**Pull request merged**